### PR TITLE
Add check up_to_date_config

### PR DESCRIPTION
**Pull request by gordonwang0**
This check verifies that the edged super config file (`/etc/aziot/edged/config.d/00-super.toml`) has a `last_modified` after `/etc/aziot/config.toml`. Prints a warning if `config.toml` was modified after `00-super.toml`, since that means the user probably forgot to run `iotedge config apply` after updating `config.toml`.

Example message:
```
‼ iotedge configuration up-to-date with config.toml - Warning
    /etc/aziot/config.toml was modified after edged's config
    You must run 'iotedge config apply' to update edged's config with the latest config.toml
```

### CODE REVIEW

File: edge-modules/edge-dashboard/src/state.rs [hunk 1]
```diff
@@ -80,9 +80,9 @@ pub fn parse_query(query: &str, pair_delim: char, kv_delim: char) -> HashMap<&st
         .collect()
 }
 
-// returns the contents of the config.yaml file from the current device
+// returns the contents of the config.toml file from the current device
```
Comment (arsing) [inline edge-modules/edge-dashboard/src/state.rs:83]: Revert the changes to edge-dashboard. It doesn't even compile; no point making more diffs in it.

File: edgelet/edgelet-core/src/error.rs [hunk 2]
```diff
@@ -73,13 +73,13 @@ pub enum ErrorKind {
     InvalidModuleType(String),
 
     #[fail(
-        display = "Error parsing URI {} specified for '{}'. Please check the config.yaml file.",
+        display = "Error parsing URI {} specified for '{}'. Please check the config.toml file.",
```
Comment (arsing) [inline edgelet/edgelet-core/src/error.rs:76]: "config.toml" isn't right either; there are multiple files. Make it "configuration" like what `aziotctl check` calls it.

Actually some of these are probably unused now, so just delete them them.

File: edgelet/iotedge/src/check/checks/connect_management_uri.rs [hunk 3]
```diff
@@ -18,7 +18,7 @@ impl Checker for ConnectManagementUri {
         "connect-management-uri"
     }
     fn description(&self) -> &'static str {
-        "config.yaml has correct URIs for daemon mgmt endpoint"
+        "config.toml has correct URIs for daemon mgmt endpoint"
```
Comment (arsing) [inline edgelet/iotedge/src/check/checks/connect_management_uri.rs:21]: Same here - "configuration"

File: edgelet/iotedge/src/check/checks/up_to_date_config.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                        .into(),
+                ))
+            }
+        };
+
+        let config_last_modified = config_metadata
+            .modified()
+            .expect("file metadata should contain valid last_modified");
+
+        for service in &["keyd", "certd", "identityd", "tpmd", "edged"] {
```
Comment (arsing) [inline edgelet/iotedge/src/check/checks/up_to_date_config.rs:48]: Implement the keyd/certd/identityd/tpmd checks in i-i-s, not here.

**Review by arsing — CHANGES_REQUESTED** [review]

File: edgelet/iotedge/src/check/checks/up_to_date_config.rs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+#[derive(Default, serde_derive::Serialize)]
+pub(crate) struct UpToDateConfig {}
+
+impl Checker for UpToDateConfig {
+    fn id(&self) -> &'static str {
+        "aziot-configs-up-to-date"
+    }
+    fn description(&self) -> &'static str {
+        "aziot configurations up-to-date with config.toml"
```
Comment (varunpuranik) [inline edgelet/iotedge/src/check/checks/up_to_date_config.rs:16]: I feel we should call it iotedge and not aziot.

File: edgelet/iotedge/src/check/checks/up_to_date_config.rs (hunk 4 above)
Comment (gordonwang0) [reply edgelet/iotedge/src/check/checks/up_to_date_config.rs:48]: So move this entire check to i-i-s? Since the keyd/certd/identityd/tpmd checks are the majority of this check.

File: edgelet/iotedge/src/check/checks/connect_management_uri.rs (hunk 3 above)
Comment (gordonwang0) [reply edgelet/iotedge/src/check/checks/connect_management_uri.rs:21]: fixed

File: edge-modules/edge-dashboard/src/state.rs (hunk 1 above)
Comment (gordonwang0) [reply edge-modules/edge-dashboard/src/state.rs:83]: fixed

File: edgelet/edgelet-core/src/error.rs (hunk 2 above)
Comment (gordonwang0) [reply edgelet/edgelet-core/src/error.rs]: fixed

File: edgelet/iotedge/src/check/checks/up_to_date_config.rs (hunk 4 above)
Comment (arsing) [reply edgelet/iotedge/src/check/checks/up_to_date_config.rs:48]: `aziotctl check` will have a check that checks the files for the i-i-s services. `iotedge check` will have a check that checks the file for edged. You can move the loop body to a function in `aziotctl-common` so that both checks can share it.

Comment (gordonwang0) [reply edgelet/iotedge/src/check/checks/up_to_date_config.rs:48]: Got it. Is there an easy way for a function in `aziotctl-common` to return the `CheckResult` type? I see that there's a different `CheckResult` for i-i-s and iotedge.

File: edgelet/iotedge/src/check/checks/up_to_date_config.rs (hunk 5 above)
Comment (gordonwang0) [reply edgelet/iotedge/src/check/checks/up_to_date_config.rs]: Fixed.

File: edgelet/iotedge/src/check/checks/up_to_date_config.rs (hunk 4 above)
Comment (arsing) [reply edgelet/iotedge/src/check/checks/up_to_date_config.rs:48]: You can return the new `CheckResult` and manually map the variants to the old `CheckResult` - `anyhow::Error` -> `failure::Error` conversion is `Context::new(format!("{:?}", err)).into()` (at the cost of flattening the context chain into a single error). It's hacky but it'll all get cleaned up when we drop `failure` anyway.

Or you can just duplicate the code instead of making a common function.

File: edgelet/iotedge/src/check/checks/container_resolve_parent_hostname.rs [hunk 6]
```diff
@@ -12,7 +12,7 @@ impl Checker for ContainerResolveParentHostname {
         "container-resolve-parent-hostname"
     }
     fn description(&self) -> &'static str {
-        "config.yaml parent hostname is resolvable from inside container"
+        "config.toml parent hostname is resolvable from inside container"
```
Comment (arsing) [inline edgelet/iotedge/src/check/checks/container_resolve_parent_hostname.rs:15]: `configuration`

File: edgelet/iotedge/src/check/checks/hostname.rs [hunk 7]
```diff
@@ -19,7 +19,7 @@ impl Checker for Hostname {
         "hostname"
     }
     fn description(&self) -> &'static str {
-        "config.yaml has correct hostname"
+        "config.toml has correct hostname"
```
Comment (arsing) [inline edgelet/iotedge/src/check/checks/hostname.rs:22]: `configuration`

File: edgelet/iotedge/src/check/checks/parent_hostname.rs [hunk 8]
```diff
@@ -16,7 +16,7 @@ impl Checker for ParentHostname {
         "parent_hostname"
     }
     fn description(&self) -> &'static str {
-        "config.yaml has correct parent_hostname"
+        "config.toml has correct parent_hostname"
```
Comment (arsing) [inline edgelet/iotedge/src/check/checks/parent_hostname.rs:19]: `configuration`

File: edgelet/iotedge/src/check/checks/up_to_date_config.rs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+#[derive(Default, serde_derive::Serialize)]
+pub(crate) struct UpToDateConfig {}
+
+impl Checker for UpToDateConfig {
+    fn id(&self) -> &'static str {
+        "iotedge-config-up-to-date"
+    }
+    fn description(&self) -> &'static str {
+        "iotedge configuration up-to-date with config.toml"
```
Comment (arsing) [inline edgelet/iotedge/src/check/checks/up_to_date_config.rs:16]: For consistency with the others, no "iotedge"

File: edgelet/iotedge/src/check/mod.rs [hunk 10]
```diff
@@ -134,7 +134,7 @@ impl Check {
                     all_checks.extend(
                         aziot_checks
                             .into_iter()
-                            .map(|(section_name, checks)| (section_name + " (aziot)", checks)),
+                            .map(|(section_name, checks)| (section_name + " (aziotctl)", checks)),
```
Comment (arsing) [inline edgelet/iotedge/src/check/mod.rs:137]: Well it's the section name rather than the binary name, so `aziot-identity-service` to indicate the package, perhaps.

File: edgelet/support-bundle/src/support_bundle.rs [hunk 11]
```diff
@@ -363,7 +363,7 @@ where
         /***
          * Note: this assumes using windows containers on a windows machine.
          * This is the expected production scenario.
-         * Since the bundle command does not read the config.yaml, it cannot use the `moby.runtime_uri` from there.
+         * Since the bundle command does not read the config.toml, it cannot use the `moby.runtime_uri` from there.
```
Comment (arsing) [inline edgelet/support-bundle/src/support_bundle.rs:366]: Just delete this whole comment block and the line of code it applies to. We don't care about `cfg(windows)` any more.

**Review by arsing — CHANGES_REQUESTED** [review]

File: edgelet/iotedge/src/check/checks/container_resolve_parent_hostname.rs (hunk 6 above)
Comment (gordonwang0) [reply edgelet/iotedge/src/check/checks/container_resolve_parent_hostname.rs:15]: Fixed

File: edgelet/iotedge/src/check/checks/hostname.rs (hunk 7 above)
Comment (gordonwang0) [reply edgelet/iotedge/src/check/checks/hostname.rs:22]: Fixed

File: edgelet/iotedge/src/check/checks/parent_hostname.rs (hunk 8 above)
Comment (gordonwang0) [reply edgelet/iotedge/src/check/checks/parent_hostname.rs:19]: Fixed

File: edgelet/iotedge/src/check/checks/up_to_date_config.rs (hunk 9 above)
Comment (gordonwang0) [reply edgelet/iotedge/src/check/checks/up_to_date_config.rs:16]: Fixed

File: edgelet/iotedge/src/check/mod.rs (hunk 10 above)
Comment (gordonwang0) [reply edgelet/iotedge/src/check/mod.rs:137]: Fixed

File: edgelet/support-bundle/src/support_bundle.rs (hunk 11 above)
Comment (gordonwang0) [reply edgelet/support-bundle/src/support_bundle.rs:366]: Fixed

File: edgelet/iotedge/src/check/checks/parent_hostname.rs (hunk 8 above)
Comment (arsing) [reply edgelet/iotedge/src/check/checks/parent_hostname.rs]: https://github.com/Azure/iotedge/pull/4596/files#diff-dcde7d0be94b057e6c90a1c65708a812d96ac53d1593f2be1f8f2ad7b2958210R57-R73

File: edgelet/iotedge/src/check/checks/hostname.rs (hunk 7 above)
Comment (arsing) [reply edgelet/iotedge/src/check/checks/hostname.rs]: https://github.com/Azure/iotedge/pull/4596/files#diff-2d2d87d941afe99ae2a3d950e0ba4acb17a01058337964c68278fc70739f14abR93-R121

File: edgelet/iotedge/src/check/mod.rs (hunk 10 above)
Comment (arsing) [reply edgelet/iotedge/src/check/mod.rs]: https://github.com/Azure/iotedge/pull/4596/files#diff-5b05b8525be1ca907a0bc313fd71312d95a964b5f60d1903aa0fc2992a247027R147 etc

File: edgelet/iotedge/src/check/checks/container_resolve_parent_hostname.rs (hunk 6 above)
Comment (arsing) [reply edgelet/iotedge/src/check/checks/container_resolve_parent_hostname.rs]: Actually, this shouldn't even have anything before `parent`

File: edgelet/iotedge/src/check/checks/hostname.rs (hunk 7 above)
Comment (arsing) [reply edgelet/iotedge/src/check/checks/hostname.rs]: Also:

https://github.com/Azure/iotedge/pull/4596/files#diff-f8e580b09869fa14b9130d7c37ee695ea83e967fefd46487f1641ea8a87a4892R16

https://github.com/Azure/iotedge/pull/4596/files#diff-5b05b8525be1ca907a0bc313fd71312d95a964b5f60d1903aa0fc2992a247027R705-R771

File: edgelet/iotedge/src/check/checks/container_resolve_parent_hostname.rs (hunk 6 above)
Comment (gordonwang0) [reply edgelet/iotedge/src/check/checks/container_resolve_parent_hostname.rs]: Fixed

File: edgelet/iotedge/src/check/checks/hostname.rs (hunk 7 above)
Comment (gordonwang0) [reply edgelet/iotedge/src/check/checks/hostname.rs]: Fixed

File: edgelet/iotedge/src/check/checks/parent_hostname.rs (hunk 8 above)
Comment (gordonwang0) [reply edgelet/iotedge/src/check/checks/parent_hostname.rs]: Fixed

File: edgelet/iotedge/src/check/mod.rs (hunk 10 above)
Comment (gordonwang0) [reply edgelet/iotedge/src/check/mod.rs]: Changed all instances of aziotctl to aziot-identity-service.

**Review by arsing — APPROVED** [review]
